File: src/main/java/org/radarcns/android/kafka/KafkaDataSubmitter.java
/*
 * Copyright 2017 The Hyve
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.radarcns.android.kafka;

import android.os.Handler;
import android.os.HandlerThread;
import android.support.annotation.NonNull;
import org.radarcns.android.data.DataCache;
import org.radarcns.android.data.DataHandler;
import org.radarcns.data.Record;
import org.radarcns.kafka.ObservationKey;
import org.radarcns.producer.AuthenticationException;
import org.radarcns.producer.KafkaSender;
import org.radarcns.producer.KafkaTopicSender;
import org.radarcns.topic.AvroTopic;
import org.radarcns.util.ListPool;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.Closeable;
import java.io.IOException;
import java.util.*;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.ConcurrentLinkedQueue;
import java.util.concurrent.ConcurrentMap;
import java.util.concurrent.atomic.AtomicInteger;

import static android.os.Process.THREAD_PRIORITY_BACKGROUND;

/**
 * Separate thread to read from the database and send it to the Kafka server. It cleans the
 * database.
 *
 * It uses a set of timers to addMeasurement data and clean the databases.
 */
public class KafkaDataSubmitter<V> implements Closeable {
    private static final Logger logger = LoggerFactory.getLogger(KafkaDataSubmitter.class);

    private final DataHandler<ObservationKey, V> dataHandler;
    private final KafkaSender<ObservationKey, V> sender;
    private final ConcurrentMap<AvroTopic<ObservationKey, V>, Queue<Record<ObservationKey, V>>> trySendCache;
    private final Map<AvroTopic<ObservationKey, V>, Runnable> trySendFuture;
    private final Map<AvroTopic<ObservationKey, V>, KafkaTopicSender<ObservationKey, V>> topicSenders;
    private final KafkaConnectionChecker connection;
    private static final ListPool listPool = new ListPool(1);
    private final AtomicInteger sendLimit;
    private final HandlerThread mHandlerThread;
    private final Handler mHandler;

    private Runnable uploadFuture;
    private Runnable uploadIfNeededFuture;
    /** Upload rate in milliseconds. */
    private long uploadRate;
    private String userId;

    public KafkaDataSubmitter(@NonNull DataHandler<ObservationKey, V> dataHandler, @NonNull
            KafkaSender<ObservationKey, V> sender, int sendLimit, long uploadRate, String userId) {
        this.dataHandler = dataHandler;
        this.sender = sender;
        this.userId = userId;
        trySendCache = new ConcurrentHashMap<>();
        trySendFuture = new HashMap<>();
        topicSenders = new HashMap<>();
        this.sendLimit = new AtomicInteger(sendLimit);

        mHandlerThread = new HandlerThread("data-submitter", THREAD_PRIORITY_BACKGROUND);
        mHandlerThread.start();
        mHandler = new Handler(mHandlerThread.getLooper());

        logger.info("Started data submission executor");

        connection = new KafkaConnectionChecker(sender, mHandler, dataHandler, uploadRate * 5);

        mHandler.post(new Runnable() {
            @Override
            public void run() {
                try {
                    if (KafkaDataSubmitter.this.sender.isConnected()) {
                        KafkaDataSubmitter.this.dataHandler.updateServerStatus(ServerStatusListener.Status.CONNECTED);
                        connection.didConnect();
                    } else {
                        KafkaDataSubmitter.this.dataHandler.updateServerStatus(ServerStatusListener.Status.DISCONNECTED);
                        connection.didDisconnect(null);
                    }
                } catch (AuthenticationException ex) {
                    connection.didDisconnect(ex);
                }
            }
        });

        synchronized (this) {
            uploadFuture = null;
            uploadIfNeededFuture = null;
            setUploadRate(uploadRate);
        }
        logger.info("Remote Config: Upload rate is '{}' sec per upload", uploadRate);
    }

    /** Set upload rate in seconds. */
    public final synchronized void setUploadRate(long period) {
        long newUploadRate = period * 1000L;
        if (this.uploadRate == newUploadRate) {
            return;
        }
        this.uploadRate = newUploadRate;
        if (uploadFuture != null) {
            mHandler.removeCallbacks(uploadFuture);
            mHandler.removeCallbacks(uploadIfNeededFuture);
        }
        // Get upload frequency from system property
        uploadFuture = new Runnable() {
            Set<AvroTopic<ObservationKey, ? extends V>> topicsToSend = Collections.emptySet();

            @Override
            public void run() {
                if (connection.isConnected()) {
                    if (topicsToSend.isEmpty()) {
                        topicsToSend = new HashSet<>(KafkaDataSubmitter.this.dataHandler.getCaches().keySet());
                    }
                    uploadCaches(topicsToSend);
                    // still more data to send, do that immediately
                    if (!topicsToSend.isEmpty()) {
                        mHandler.post(this);
                        return;
                    }
                } else {
                    topicsToSend.clear();
                }
                mHandler.postDelayed(this, uploadRate);
            }
        };
        mHandler.postDelayed(uploadFuture, uploadRate);

        uploadIfNeededFuture = new Runnable() {
            @Override
            public void run() {
                if (connection.isConnected()) {
                    boolean sendAgain = uploadCachesIfNeeded();
                    if (sendAgain) {
                        mHandler.post(this);
                        return;
                    }
                }
                mHandler.postDelayed(this, uploadRate / 5);
            }
        };
        mHandler.postDelayed(uploadIfNeededFuture, uploadRate / 5);
    }

    /** Upload rate in seconds. */
    private synchronized long getUploadRate() {
        return this.uploadRate / 1000L;
    }

    public void setSendLimit(int limit) {
        sendLimit.set(limit);
    }

    /**
     * Close the submitter eventually. This does not flush any caches.
     */
    @Override
    public synchronized void close() {
        mHandler.post(new Runnable() {
            @Override
            public void run() {
                mHandler.removeCallbacks(uploadFuture);
                mHandler.removeCallbacks(uploadIfNeededFuture);

                synchronized (trySendFuture) {
                    for (Runnable future : trySendFuture.values()) {
                        mHandler.removeCallbacks(future);
                    }
                    trySendFuture.clear();
                    trySendCache.clear();
                }

                for (Map.Entry<AvroTopic<ObservationKey, V>, KafkaTopicSender<ObservationKey, V>> topicSender : topicSenders.entrySet()) {
                    try {
                        topicSender.getValue().close();
                    } catch (IOException e) {
                        logger.warn("failed to close topicSender for topic {}", topicSender.getKey().getName(), e);
                    }
                }
                topicSenders.clear();

                try {
                    sender.close();
                } catch (IOException e1) {
                    logger.warn("failed to addMeasurement latest batches", e1);
                }
            }
        });
        mHandlerThread.quitSafely();
    }

    /** Get a sender for a topic. Per topic, only ONE thread may use this. */
    private KafkaTopicSender<ObservationKey, V> sender(AvroTopic<ObservationKey, V> topic) throws IOException {
        KafkaTopicSender<ObservationKey, V> topicSender = topicSenders.get(topic);
        if (topicSender == null) {
            topicSender = sender.sender(topic);
            topicSenders.put(topic, topicSender);
        }
        return topicSender;
    }

    /**
     * Check the connection status eventually.
     */
    public void checkConnection() {
        connection.check();
    }

    /**
     * Upload the caches if they would cause the buffer to overflow
     */
    private boolean uploadCachesIfNeeded() {
        boolean uploadingNotified = false;
        int currentSendLimit = sendLimit.get();
        boolean sendAgain = false;

        try {
            for (Map.Entry<AvroTopic<ObservationKey, ? extends V>, ? extends DataCache<ObservationKey, ? extends V>> entry
                    : dataHandler.getCaches().entrySet()) {
                long unsent = entry.getValue().numberOfRecords().first;
                if (unsent > currentSendLimit) {
                    //noinspection unchecked
                    int sent = uploadCache(
                            (AvroTopic<ObservationKey, V>) entry.getKey(), (DataCache<ObservationKey, V>) entry.getValue(),
                            currentSendLimit, uploadingNotified);
                    if (!uploadingNotified) {
                        uploadingNotified = true;
                    }
                    if (!sendAgain && unsent - sent > currentSendLimit && sent != -1) {
                        sendAgain = true;
                    }
                }
            }
            if (uploadingNotified) {
                dataHandler.updateServerStatus(ServerStatusListener.Status.CONNECTED);
                connection.didConnect();
            }
        } catch (IOException ex) {
            connection.didDisconnect(ex);
            sendAgain = false;
        }
        return sendAgain;
    }

    /**
     * Upload a limited amount of data stored in the database which is not yet sent.
     */
    private void uploadCaches(Set<AvroTopic<ObservationKey, ? extends V>> toSend) {
        boolean uploadingNotified = false;
        int currentSendLimit = sendLimit.get();
        try {
            for (Map.Entry<AvroTopic<ObservationKey, ? extends V>, ? extends DataCache<ObservationKey, ? extends V>> entry : dataHandler.getCaches().entrySet()) {
                if (!toSend.contains(entry.getKey())) {
                    continue;
                }
                @SuppressWarnings("unchecked") // we can upload any record
                int sent = uploadCache((AvroTopic<ObservationKey, V>)entry.getKey(), (DataCache<ObservationKey, V>)entry.getValue(), currentSendLimit, uploadingNotified);
                if (sent < currentSendLimit) {
                    toSend.remove(entry.getKey());
                }
                if (!uploadingNotified && sent > 0) {
                    uploadingNotified = true;
                }
            }
            if (uploadingNotified) {
                dataHandler.updateServerStatus(ServerStatusListener.Status.CONNECTED);
                connection.didConnect();
            }
        } catch (IOException ex) {
            connection.didDisconnect(ex);
        }
    }

    /**
     * Upload some data from a single table.
     * @return number of records sent.
     */
    private int uploadCache(AvroTopic<ObservationKey, V> topic, DataCache<ObservationKey, V> cache, int limit,
                            boolean uploadingNotified) throws IOException {
        List<Record<ObservationKey, V>> unfilteredMeasurements = cache.unsentRecords(limit);

        List<Record<ObservationKey, V>> measurements = listPool.get(Collections
                .<Record<ObservationKey,V>>emptyList());
        for (Record<ObservationKey, V> record : unfilteredMeasurements) {
            if (record.key.getUserId().equals(userId)) {
                measurements.add(record);
            }
        }

        int numberOfRecords = measurements.size();
        int totalSize = unfilteredMeasurements.size();

        if (numberOfRecords > 0) {
            KafkaTopicSender<ObservationKey, V> cacheSender = sender(topic);

            if (!uploadingNotified) {
                dataHandler.updateServerStatus(ServerStatusListener.Status.UPLOADING);
            }

            try {
                cacheSender.send(measurements);
                cacheSender.flush();
            } catch (AuthenticationException ex) {
                dataHandler.updateRecordsSent(topic.getName(), -1);
                throw ex;
            } catch (IOException ioe) {
                dataHandler.updateServerStatus(ServerStatusListener.Status.UPLOADING_FAILED);
                dataHandler.updateRecordsSent(topic.getName(), -1);
                throw ioe;
            }

            long lastOffsetSent = cacheSender.getLastSentOffset();
            cache.markSent(lastOffsetSent);

            dataHandler.updateRecordsSent(topic.getName(), numberOfRecords);

            logger.debug("uploaded {} {} records", numberOfRecords, topic.getName());

            long lastOffsetPut = measurements.get(numberOfRecords - 1).offset;
            if (lastOffsetSent == lastOffsetPut) {
                if (totalSize > numberOfRecords) {
                    long offsetGot = unfilteredMeasurements.get(totalSize - 1).offset;
                    if (offsetGot > lastOffsetPut) {
                        cache.markSent(offsetGot);
                    }
                }
                listPool.add(measurements);
            }
        } else if (totalSize > 0) {
            cache.markSent(unfilteredMeasurements.get(totalSize - 1).offset);
            listPool.add(measurements);
        } else {
            listPool.add(measurements);
        }
        cache.returnList(unfilteredMeasurements);

        return totalSize;
    }

    /**
     * Try to addMeasurement a message, without putting it in any permanent storage. Any failure may cause
     * messages to be lost. If the sender is disconnected, messages are immediately discarded.
     * @return whether the message was queued for sending.
     */
    public <W extends V> boolean trySend(final AvroTopic<ObservationKey, W> topic, final long offset,
                                         final ObservationKey deviceId, final W record) {
        if (!connection.isConnected()) {
            return false;
        }
        @SuppressWarnings("unchecked")
        final AvroTopic<ObservationKey, V> castTopic = (AvroTopic<ObservationKey, V>)topic;
        Queue<Record<ObservationKey, V>> records = trySendCache.get(castTopic);
        if (records == null) {
            records = new ConcurrentLinkedQueue<>();
            //noinspection unchecked
            trySendCache.put((AvroTopic<ObservationKey, V>)topic, records);
        }
        records.add(new Record<>(offset, deviceId, (V)record));

        long period = getUploadRate();

        synchronized (trySendFuture) {
            if (!trySendFuture.containsKey(topic)) {
                Runnable future = new Runnable() {
                    @Override
                    public void run() {
                        if (!connection.isConnected()) {
                            return;
                        }

                        List<Record<ObservationKey, V>> localRecords;

                        synchronized (trySendFuture) {
                            Queue<Record<ObservationKey, V>> queue = trySendCache.get(topic);
                            trySendFuture.remove(topic);
                            localRecords = listPool.get(queue);
                        }

                        try {
                            doImmediateSend(castTopic, localRecords);
                            connection.didConnect();
                        } catch (IOException e) {
                            dataHandler.updateRecordsSent(topic.getName(), -1);
                            connection.didDisconnect(e);
                        }
                    }
                };
                mHandler.postDelayed(future, period * 1000L);
                trySendFuture.put(castTopic, future);
            }
        }
        return true;
    }

    /** Immediately send given records, without any error recovery. */
    private void doImmediateSend(AvroTopic<ObservationKey, V> topic, List<Record<ObservationKey, V>> records)
            throws IOException {
        KafkaTopicSender<ObservationKey, V> localSender = sender(topic);
        localSender.send(records);
        dataHandler.updateRecordsSent(topic.getName(), records.size());
        long lastOffset = records.get(records.size() - 1).offset;
        if (localSender.getLastSentOffset() == lastOffset) {
            listPool.add(records);
        }
    }

    public void setUserId(String userId) {
        this.userId = userId;
    }
}

File: radar-commons-android/src/main/java/org/radarbase/android/kafka/KafkaDataSubmitter.kt
/*
 * Copyright 2017 The Hyve
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.radarbase.android.kafka

import android.os.Process
import org.apache.avro.Schema
import org.apache.avro.SchemaValidationException
import org.apache.avro.generic.IndexedRecord
import org.radarbase.android.RadarService
import org.radarbase.android.data.DataCacheGroup
import org.radarbase.android.data.DataHandler
import org.radarbase.android.data.ReadableDataCache
import org.radarbase.android.source.PluginMetadataStore
import org.radarbase.android.util.FirebaseEventLogger
import org.radarbase.android.util.SafeHandler
import org.radarbase.data.AvroRecordData
import org.radarbase.data.RecordData
import org.radarbase.producer.AuthenticationException
import org.radarbase.producer.KafkaSender
import org.radarbase.producer.KafkaTopicSender
import org.radarbase.topic.AvroTopic
import org.slf4j.LoggerFactory
import java.io.Closeable
import java.io.IOException
import java.util.*
import java.util.concurrent.atomic.AtomicBoolean
import kotlin.collections.HashSet

/**
 * Separate thread to read from the database and send it to the Kafka server. It cleans the
 * database.
 *
 * It uses a set of timers to addMeasurement data and clean the databases.
 */
class KafkaDataSubmitter(
    private val dataHandler: DataHandler<*, *>,
    private val sender: KafkaSender,
    config: SubmitterConfiguration,
    radarService: RadarService? = null,
) : Closeable {

    private val submitHandler = SafeHandler.getInstance("KafkaDataSubmitter", Process.THREAD_PRIORITY_BACKGROUND)
    private val topicSenders: MutableMap<String, KafkaTopicSender<Any, Any>> = HashMap()
    private val connection: KafkaConnectionChecker
    private val pluginMetadata: PluginMetadataStore? = radarService?.pluginMetadata

    var config: SubmitterConfiguration = config
        set(newValue) {
            this.submitHandler.execute {
                if (newValue == field) return@execute

                validate(newValue)
                field = newValue.copy()
                schedule()
            }
        }

    private var uploadFuture: SafeHandler.HandlerFuture? = null
    private var uploadIfNeededFuture: SafeHandler.HandlerFuture? = null
    /** Upload rate in milliseconds.  */

    init {
        validate(config)

        submitHandler.start()

        logger.debug("Started data submission executor")

        connection = KafkaConnectionChecker(sender, submitHandler, dataHandler, config.uploadRate * 5)

        submitHandler.execute {
            uploadFuture = null
            uploadIfNeededFuture = null

            try {
                if (sender.isConnected) {
                    dataHandler.updateServerStatus(ServerStatusListener.Status.CONNECTED)
                    connection.didConnect()
                } else {
                    dataHandler.updateServerStatus(ServerStatusListener.Status.DISCONNECTED)
                    connection.didDisconnect(null)
                }
            } catch (ex: AuthenticationException) {
                connection.didDisconnect(ex)
            }

            schedule()
        }
    }

    private fun validate(config: SubmitterConfiguration) {
        requireNotNull(config.userId) { "User ID is mandatory to start KafkaDataSubmitter" }
    }

    private fun schedule() {
        val uploadRate = config.uploadRate * config.uploadRateMultiplier * 1000L
        uploadFuture?.cancel()
        uploadIfNeededFuture?.cancel()

        // Get upload frequency from system property
        uploadFuture = this.submitHandler.repeat(uploadRate) {
            val topicsToSend = dataHandler.activeCaches.mapTo(HashSet()) { it.topicName }
            while (connection.isConnected && topicsToSend.isNotEmpty()) {
                logger.debug("Uploading topics {}", topicsToSend)
                uploadCaches(topicsToSend)
            }
        }

        uploadIfNeededFuture = this.submitHandler.repeat(uploadRate / 5) {
            var sendAgain = true
            while (connection.isConnected && sendAgain) {
                logger.debug("Uploading full topics")
                sendAgain = uploadCachesIfNeeded()
            }
        }
    }

    /**
     * Close the submitter eventually. This does not flush any caches.
     */
    @Synchronized
    override fun close() {
        this.submitHandler.stop {
            for ((topic, sender) in topicSenders) {
                try {
                    sender.close()
                } catch (e: IOException) {
                    logger.warn("failed to stop topicSender for topic {}", topic, e)
                }
            }
            topicSenders.clear()

            try {
                sender.close()
            } catch (e1: IOException) {
                logger.warn("failed to addMeasurement latest batches", e1)
            }
        }
    }

    /** Get a sender for a topic. Per topic, only ONE thread may use this.  */
    @Throws(IOException::class, SchemaValidationException::class)
    private fun sender(
        topic: AvroTopic<Any, Any>,
    ): KafkaTopicSender<Any, Any> = topicSenders.computeIfAbsentKt(topic.name) { sender.sender(topic) }

    private fun <K: Any, V: Any> MutableMap<K, V>.computeIfAbsentKt(
        key: K,
        mappingFunction: (K) -> V,
    ): V = get(key)
        ?: mappingFunction(key).also { put(key, it) }

    /**
     * Check the connection status eventually.
     */
    fun checkConnection() {
        connection.check()
    }

    /**
     * Upload the caches if they would cause the buffer to overflow
     */
    private fun uploadCachesIfNeeded(): Boolean {
        var sendAgain = false

        try {
            val uploadingNotified = AtomicBoolean(false)

            for (entry in dataHandler.activeCaches) {
                val unsent = entry.activeDataCache.numberOfRecords
                if (unsent > config.amountLimit) {
                    val sent = uploadCache(entry.activeDataCache, uploadingNotified)
                    if (unsent - sent > config.amountLimit) {
                        sendAgain = true
                    }
                }
            }
            if (uploadingNotified.get()) {
                dataHandler.updateServerStatus(ServerStatusListener.Status.CONNECTED)
                connection.didConnect()
            }
        } catch (ex: Exception) {
            connection.didDisconnect(ex)
            sendAgain = false
        }

        return sendAgain
    }

    /**
     * Upload a limited amount of data stored in the database which is not yet sent.
     */
    private fun uploadCaches(toSend: MutableSet<String>) {
        try {
            val uploadingNotified = AtomicBoolean(false)
            toSend -= dataHandler.activeCaches
                .asSequence()
                .filter { group ->
                    if (group.topicName in toSend) {
                        val sentActive = uploadCache(group.activeDataCache, uploadingNotified)
                        val sentDeprecated = group.deprecatedCaches.map { uploadCache(it, uploadingNotified) }

                        if (sentDeprecated.any { it == 0 }) {
                            group.deleteEmptyCaches()
                        }
                        sentActive < config.amountLimit
                                && sentDeprecated.all { it < config.amountLimit }
                    } else false
                }
                .mapTo(HashSet(), DataCacheGroup<*,*>::topicName)

            if (uploadingNotified.get()) {
                dataHandler.updateServerStatus(ServerStatusListener.Status.CONNECTED)
                connection.didConnect()
            }
        } catch (ex: Exception) {
            connection.didDisconnect(ex)
        }
    }

    /**
     * Upload some data from a single table.
     * @return number of records sent.
     */
    @Throws(IOException::class, SchemaValidationException::class)
    private fun uploadCache(cache: ReadableDataCache, uploadingNotified: AtomicBoolean): Int {
        val data = cache.getUnsentRecords(config.amountLimit, config.sizeLimit)
            ?: return 0

        val size = data.size()
        if (size == 0) {
            return 0
        }

        val recordsNotNull = data.filterNotNull()

        if (recordsNotNull.isNotEmpty()) {
            val topic = cache.readTopic

            val keyUserId: String? = if (topic.keySchema.type == Schema.Type.RECORD) {
                topic.keySchema.getField("userId")?.let { userIdField ->
                    (data.key as IndexedRecord).get(userIdField.pos()).toString()
                }
            } else null

            val keyProjectId: String? = retrieveDataFromFields(topic, data, "projectId") as? String

            if ((keyUserId == null || keyUserId == config.userId) && (keyProjectId == null || keyProjectId == config.projectId)) {
                if (uploadingNotified.compareAndSet(false, true)) {
                    dataHandler.updateServerStatus(ServerStatusListener.Status.UPLOADING)
                }
                val keySourceId = retrieveDataFromFields(topic, data, "sourceId") as? String
                if (pluginMetadata != null) {
                    if (keySourceId != null && keySourceId !in pluginMetadata.sourceIds) {
                        logger.warn(
                            "(MismatchedId) SourceId: {} for topic: {} doesn't match with any sourceId's. Discarding the data",
                            keySourceId,
                            topic.name
                        )
                        dataHandler.let {
                            it.updateRecordsSent(topic.name, size.toLong())
                            // The upload has not actually failed yet, but if it proceeds, it will fail due to an incorrect sourceId.
                            // Therefore, the status is explicitly set to UPLOADING_FAILED.
                            it.updateServerStatus(ServerStatusListener.Status.UPLOADING_FAILED)
                        }

                        cache.remove(size)

                        val pluginName = pluginMetadata.topicToPluginMap.get(topic.name)
                        val sourceId = pluginMetadata.pluginToSourceIdMap[pluginName]

                        FirebaseEventLogger.reportMismatchedSourceId(
                            keyProjectId,
                            keySourceId,
                            keyUserId,
                            topic.name,
                            pluginName,
                            sourceId
                        )

                        return size
                    }
                }
                try {
                    sender(topic).run {
                        send(AvroRecordData<Any, Any>(data.topic, data.key, recordsNotNull))
                        flush()
                    }
                    dataHandler.updateRecordsSent(topic.name, size.toLong())
                } catch (ex: AuthenticationException) {
                    dataHandler.updateRecordsSent(topic.name, -1)
                    throw ex
                } catch (e: Exception) {
                    dataHandler.updateServerStatus(ServerStatusListener.Status.UPLOADING_FAILED)
                    dataHandler.updateRecordsSent(topic.name, -1)
                    throw e
                }

                logger.debug("uploaded {} {} records", size, topic.name)
            } else {
                val pluginName = pluginMetadata?.topicToPluginMap?.get(topic.name)

                when {
                    keyProjectId != config.projectId -> {
                        FirebaseEventLogger.reportMismatchedProjectId(
                            keyProjectId, config.projectId, pluginName, topic.name, keyUserId
                        )
                    }

                    keyUserId != config.userId -> {
                        FirebaseEventLogger.reportMismatchedUserId(
                            keyUserId, config.userId, pluginName, topic.name
                        )
                    }
                }
            }
        }

        cache.remove(size)
        return size
    }

    private fun retrieveDataFromFields(topic: AvroTopic<Any, Any>, data: RecordData<Any, Any?>, field: String) =
        if (topic.keySchema.type == Schema.Type.RECORD) {
            topic.keySchema.getField(field)?.let { fieldName ->
                (data.key as IndexedRecord).get(fieldName.pos()).toString()
            }
        } else null

    companion object {
        private val logger = LoggerFactory.getLogger(KafkaDataSubmitter::class.java)
    }
}
